File: src/main/java/ch/epfl/dias/calcite/adapter/pelago/types/PelagoTypeParser.java
package ch.epfl.dias.calcite.adapter.pelago.types;

import org.apache.calcite.linq4j.tree.Primitive;
import org.apache.calcite.rel.type.RelDataType;
import org.apache.calcite.rel.type.RelDataTypeFactory;
import org.apache.calcite.sql.type.SqlTypeName;
import org.apache.calcite.util.Pair;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;

public class PelagoTypeParser {

    public static RelDataType parseType(RelDataTypeFactory typeFactory, Map<String, ?> type) throws IOException {
        switch ((String) type.getOrDefault("type", null)){
            case "int":
                return parseInt(typeFactory, type);
            case "int64":
                return parseInt64(typeFactory, type);
            case "float":
                return parseFloat(typeFactory, type);
            case "bool":
                return parseBoolean(typeFactory, type);
            case "dstring":
                return parseDString(typeFactory, type);
            case "date":
                return parseDate(typeFactory, type);
            case "string":
                return parseString(typeFactory, type);
            case "set":
                return parseSet(typeFactory, type);
            case "bag":
                return parseBag(typeFactory, type);
            case "list":
                return parseList(typeFactory, type);
            case "record":
                return parseRecord(typeFactory, type);
            default:
                throw new IOException("unknown type: " + type);
        }
    }

    public static RelDataType parseInt(RelDataTypeFactory typeFactory, Map<String, ?> type){
        assert(type.getOrDefault("type", null).equals("int"));
        RelDataType javaType = typeFactory.createJavaType(Primitive.INT.boxClass);
        RelDataType sqlType = typeFactory.createSqlType(javaType.getSqlTypeName());
        return typeFactory.createTypeWithNullability(sqlType, true);
    }

    public static RelDataType parseInt64(RelDataTypeFactory typeFactory, Map<String, ?> type){
        assert(type.getOrDefault("type", null).equals("int64"));
        RelDataType javaType = typeFactory.createJavaType(Primitive.LONG.boxClass);
        RelDataType sqlType = typeFactory.createSqlType(javaType.getSqlTypeName());
        return typeFactory.createTypeWithNullability(sqlType, true);
    }

    public static RelDataType parseFloat(RelDataTypeFactory typeFactory, Map<String, ?> type){
        assert(type.getOrDefault("type", null).equals("float"));
        RelDataType javaType = typeFactory.createJavaType(Primitive.DOUBLE.boxClass);
        RelDataType sqlType = typeFactory.createSqlType(javaType.getSqlTypeName());
        return typeFactory.createTypeWithNullability(sqlType, true);
    }

    public static RelDataType parseDate(RelDataTypeFactory typeFactory, Map<String, ?> type){
        assert(type.getOrDefault("type", null).equals("date"));
        RelDataType sqlType = typeFactory.createSqlType(SqlTypeName.DATE);
        return typeFactory.createTypeWithNullability(sqlType, true);
    }

    public static RelDataType parseBoolean(RelDataTypeFactory typeFactory, Map<String, ?> type){
        assert(type.getOrDefault("type", null).equals("bool"));
        RelDataType javaType = typeFactory.createJavaType(Primitive.BOOLEAN.boxClass);
        RelDataType sqlType = typeFactory.createSqlType(javaType.getSqlTypeName());
        return typeFactory.createTypeWithNullability(sqlType, true);
    }

    public static RelDataType parseDString(RelDataTypeFactory typeFactory, Map<String, ?> type){
        assert(type.getOrDefault("type", null).equals("dstring"));
        RelDataType javaType = typeFactory.createJavaType(String.class);
        RelDataType sqlType = typeFactory.createSqlType(SqlTypeName.VARCHAR);
        return typeFactory.createTypeWithNullability(sqlType, true);
    }

    public static RelDataType parseString(RelDataTypeFactory typeFactory, Map<String, ?> type){
        assert(type.getOrDefault("type", null).equals("string"));
        RelDataType javaType = typeFactory.createJavaType(String.class);
        RelDataType sqlType = typeFactory.createSqlType(SqlTypeName.VARCHAR);
        return typeFactory.createTypeWithNullability(sqlType, true);
    }

    public static RelDataType parseSet(RelDataTypeFactory typeFactory, Map<String, ?> type) throws IOException {
        assert(type.getOrDefault("type", null).equals("set"));
        RelDataType innerType = parseType(typeFactory, (Map<String, ?>) type.get("inner"));
        RelDataType sqlType = typeFactory.createMultisetType(innerType, -1); //TODO: not really a multiset
        return typeFactory.createTypeWithNullability(sqlType, true);
    }

    public static RelDataType parseBag(RelDataTypeFactory typeFactory, Map<String, ?> type) throws IOException {
        assert(type.getOrDefault("type", null).equals("bag"));
        RelDataType innerType = parseType(typeFactory, (Map<String, ?>) type.get("inner"));
        RelDataType sqlType = typeFactory.createMultisetType(innerType, -1);
        return typeFactory.createTypeWithNullability(sqlType, true);
    }

    public static RelDataType parseList(RelDataTypeFactory typeFactory, Map<String, ?> type) throws IOException {
        assert(type.getOrDefault("type", null).equals("list"));
        RelDataType innerType = parseType(typeFactory, (Map<String, ?>) type.get("inner"));
        RelDataType sqlType = typeFactory.createArrayType(innerType, -1);
        return typeFactory.createTypeWithNullability(sqlType, true);
    }

    public static RelDataType parseRecord(RelDataTypeFactory typeFactory, Map<String, ?> type) throws IOException {
        assert(type.getOrDefault("type", null).equals("record"));
        List<Map<String, ?>> attributes = (List<Map<String, ?>>) type.get("attributes");
        List<RelDataType> types = new ArrayList<RelDataType>();
        List<String     > names = new ArrayList<String     >();
        for (Map<String, ?> attr: attributes){
            types.add(parseType(typeFactory, (Map<String, ?>) attr.get("type")));
            names.add((String) attr.get("attrName"));
        }
        return typeFactory.createStructType(Pair.zip(names, types));
    }
}

File: core/planner/src/main/scala/ch/epfl/dias/calcite/adapter/pelago/types/PelagoTypeParser.scala
/*
    Proteus -- High-performance query processing on heterogeneous hardware.

                            Copyright (c) 2017
        Data Intensive Applications and Systems Laboratory (DIAS)
                École Polytechnique Fédérale de Lausanne

                            All Rights Reserved.

    Permission to use, copy, modify and distribute this software and
    its documentation is hereby granted, provided that both the
    copyright notice and this permission notice appear in all copies of
    the software, derivative works or modified versions, and any
    portions thereof, and that both notices appear in supporting
    documentation.

    This code is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. THE AUTHORS
    DISCLAIM ANY LIABILITY OF ANY KIND FOR ANY DAMAGES WHATSOEVER
    RESULTING FROM THE USE OF THIS SOFTWARE.
 */

package ch.epfl.dias.calcite.adapter.pelago.types

import org.apache.calcite.linq4j.tree.Primitive
import org.apache.calcite.rel.`type`.{RelDataType, RelDataTypeFactory}
import org.apache.calcite.sql.`type`.SqlTypeName
import org.apache.calcite.util.Pair

import java.io.IOException
import java.util
import scala.collection.JavaConverters._

object PelagoTypeParser {

  @throws[IOException]
  def parseType(
      typeFactory: RelDataTypeFactory,
      `type`: util.Map[String, Any]
  ): RelDataType =
    `type`
      .getOrDefault("type", null)
      .asInstanceOf[String] match {
      case "int" =>
        parseInt(typeFactory, `type`)
      case "int64" =>
        parseInt64(typeFactory, `type`)
      case "float" =>
        parseFloat(typeFactory, `type`)
      case "bool" =>
        parseBoolean(typeFactory, `type`)
      case "dstring" =>
        parseDString(typeFactory, `type`)
      case "date" =>
        parseDate(typeFactory, `type`)
      case "datetime" =>
        parseDatetime(typeFactory, `type`)
      case "string" =>
        parseString(typeFactory, `type`)
      case "set" =>
        parseSet(typeFactory, `type`)
      case "bag" =>
        parseBag(typeFactory, `type`)
      case "list" =>
        parseList(typeFactory, `type`)
      case "record" =>
        parseRecord(typeFactory, `type`)
      case _ =>
        throw new IOException("unknown type: " + `type`)
    }

  def parseInt(
      typeFactory: RelDataTypeFactory,
      `type`: util.Map[String, Any]
  ): RelDataType = {
    assert(`type`.getOrDefault("type", null) == "int")
    val javaType = typeFactory.createJavaType(Primitive.INT.boxClass)
    val sqlType = typeFactory.createSqlType(javaType.getSqlTypeName)
    typeFactory.createTypeWithNullability(sqlType, true)
  }

  def parseInt64(
      typeFactory: RelDataTypeFactory,
      `type`: util.Map[String, Any]
  ): RelDataType = {
    assert(`type`.getOrDefault("type", null) == "int64")
    val javaType = typeFactory.createJavaType(Primitive.LONG.boxClass)
    val sqlType = typeFactory.createSqlType(javaType.getSqlTypeName)
    typeFactory.createTypeWithNullability(sqlType, true)
  }

  def parseFloat(
      typeFactory: RelDataTypeFactory,
      `type`: util.Map[String, Any]
  ): RelDataType = {
    assert(`type`.getOrDefault("type", null) == "float")
    val javaType = typeFactory.createJavaType(Primitive.DOUBLE.boxClass)
    val sqlType = typeFactory.createSqlType(javaType.getSqlTypeName)
    typeFactory.createTypeWithNullability(sqlType, true)
  }

  def parseDate(
      typeFactory: RelDataTypeFactory,
      `type`: util.Map[String, Any]
  ): RelDataType = {
    assert(`type`.getOrDefault("type", null) == "date")
    val sqlType = typeFactory.createSqlType(SqlTypeName.DATE)
    typeFactory.createTypeWithNullability(sqlType, true)
  }

  def parseDatetime(
      typeFactory: RelDataTypeFactory,
      `type`: util.Map[String, Any]
  ): RelDataType = {
    assert(`type`.getOrDefault("type", null) == "datetime")
    val sqlType = typeFactory.createSqlType(SqlTypeName.TIMESTAMP)
    typeFactory.createTypeWithNullability(sqlType, true)
  }

  def parseBoolean(
      typeFactory: RelDataTypeFactory,
      `type`: util.Map[String, Any]
  ): RelDataType = {
    assert(`type`.getOrDefault("type", null) == "bool")
    val javaType = typeFactory.createJavaType(Primitive.BOOLEAN.boxClass)
    val sqlType = typeFactory.createSqlType(javaType.getSqlTypeName)
    typeFactory.createTypeWithNullability(sqlType, true)
  }

  def parseDString(
      typeFactory: RelDataTypeFactory,
      `type`: util.Map[String, Any]
  ): RelDataType = {
    assert(`type`.getOrDefault("type", null) == "dstring")
    val sqlType = typeFactory.createSqlType(SqlTypeName.VARCHAR)
    typeFactory.createTypeWithNullability(sqlType, true)
  }

  def parseString(
      typeFactory: RelDataTypeFactory,
      `type`: util.Map[String, Any]
  ): RelDataType = {
    assert(`type`.getOrDefault("type", null) == "string")
    val sqlType = typeFactory.createSqlType(SqlTypeName.VARCHAR)
    typeFactory.createTypeWithNullability(sqlType, true)
  }

  @throws[IOException]
  def parseSet(
      typeFactory: RelDataTypeFactory,
      `type`: util.Map[String, Any]
  ): RelDataType = {
    assert(`type`.getOrDefault("type", null) == "set")
    val innerType = parseType(
      typeFactory,
      `type`.get("inner").asInstanceOf[util.Map[String, Any]]
    )
    val sqlType =
      typeFactory.createMultisetType(
        innerType,
        -1
      ) //TODO: not really a multiset
    typeFactory.createTypeWithNullability(sqlType, true)
  }

  @throws[IOException]
  def parseBag(
      typeFactory: RelDataTypeFactory,
      `type`: util.Map[String, Any]
  ): RelDataType = {
    assert(`type`.getOrDefault("type", null) == "bag")
    val innerType = parseType(
      typeFactory,
      `type`.get("inner").asInstanceOf[util.Map[String, Any]]
    )
    val sqlType = typeFactory.createMultisetType(innerType, -1)
    typeFactory.createTypeWithNullability(sqlType, true)
  }

  @throws[IOException]
  def parseList(
      typeFactory: RelDataTypeFactory,
      `type`: util.Map[String, Any]
  ): RelDataType = {
    assert(`type`.getOrDefault("type", null) == "list")
    val innerType = parseType(
      typeFactory,
      `type`.get("inner").asInstanceOf[util.Map[String, Any]]
    )
    val sqlType = typeFactory.createArrayType(innerType, -1)
    typeFactory.createTypeWithNullability(sqlType, true)
  }

  @throws[IOException]
  def parseRecord(
      typeFactory: RelDataTypeFactory,
      `type`: util.Map[String, Any]
  ): RelDataType = {
    assert(`type`.getOrDefault("type", null) == "record")
    val attributes =
      `type`.get("attributes").asInstanceOf[util.List[util.Map[String, _]]]
    val types = new util.ArrayList[RelDataType]
    val names = new util.ArrayList[String]
    for (attr <- attributes.asScala) {
      types.add(
        parseType(
          typeFactory,
          attr.get("type").asInstanceOf[util.Map[String, Any]]
        )
      )
      names.add(attr.get("attrName").asInstanceOf[String])
    }
    typeFactory.createStructType(Pair.zip(names, types))
  }
}
